src/HookListener/Registration.php

```diff
@@ -0,0 +1,155 @@
+     * @param Connection               $connection      The database connection.
+     * @param LoggerInterface          $logger          The logger.
+     * @param EventDispatcherInterface $eventDispatcher The event dispatcher.
+     * @param RequestStack             $requestStack    The request stack.
+     */
+    public function __construct(UserProviderInterface $userProvider, TokenStorageInterface $tokenStorage, Connection $connection, LoggerInterface $logger, EventDispatcherInterface $eventDispatcher, RequestStack $requestStack) {
+        $this->userProvider    = $userProvider;
+        $this->tokenStorage    = $tokenStorage;
+        $this->connection      = $connection;
+        $this->logger          = $logger;
+        $this->eventDispatcher = $eventDispatcher;
+        $this->requestStack    = $requestStack;
+    }
+
+    /**
+     * Within the registration process, log in the user if needed.
+     *
+     * @param int   $userId The user id
+     * @param array $data   The user data of the registration module
+     */
+    public function onCreateNewUser(int $userId, array $data): void
+    {
+        global $objPage;
+
+        $statement = $this->connection->createQueryBuilder()
+            ->select('*')
+            ->from('tl_page')
+            ->where('id=:id')
+            ->setParameter('id', $objPage->rootId)
+            ->execute();
+
+        $result = $statement->fetch(\PDO::FETCH_OBJ);
+        if (false === $result) {
+            return;
+        }
+
+        if ($result->auto_activate_registration) {
+            $match = $this->connection->createQueryBuilder()
+                ->update('tl_member')
+                ->set('disable', '')
+                ->where('id=:id')
+                ->setParameter('id', $userId)
+                ->execute();
+
+            // TODO support where
+
+            if ($result->auto_login_registration && $match) {
+                $this->loginUser($data['username']);
+            }
+        }
+    }
+
+    /**
+     * Within the activation process, log in the user if needed.
+     *
+     * @param MemberModel $member
+     */
+    public function onActivateAccount(MemberModel $member): void
+    {
+        global $objPage;
+
+        $statement = $this->connection->createQueryBuilder()
+            ->select('*')
+            ->from('tl_page')
+            ->where('id=:id')
+            ->setParameter('id', $objPage->rootId)
+            ->execute();
+
+        $result = $statement->fetch(\PDO::FETCH_OBJ);
+        if (false === $result) {
+            return;
+        }
+
+        if ($result->auto_login_activation) {
+            $this->loginUser($member->username);
+        }
+    }
+
+    /**
+     * Actually log in the user by given username.
+     *
+     * @param string $username
+     */
+    private function loginUser(string $username): void
+    {
+        // Authenticate user
+        $user = $this->userProvider->loadUserByUsername($username);
```

Comment:
Pretty sure you should try-catch here and also check if `$user instanceof \Contao\FrontendUser`